lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -0,0 +1,131 @@
+"""
+Content Library Transformer, used to filter course structure per user.
+"""
+import json
+from courseware.access import _has_access_to_course
+from courseware.models import StudentModule
+from opaque_keys.edx.locator import BlockUsageLocator
+from openedx.core.lib.block_cache.transformer import BlockStructureTransformer
+from xmodule.library_content_module import LibraryContentModule
+from xmodule.library_tools import LibraryToolsService
+from xmodule.modulestore.django import modulestore
+from eventtracking import tracker
+
+
+class ContentLibraryTransformer(BlockStructureTransformer):
+    """
+    Content Library Transformer Class
+    """
+    VERSION = 1
+
+    @classmethod
+    def _get_selected_modules(cls, user, course_key, block_key):
+        """
+        Get list of selected modules in a library,
+        for user.
+
+        Arguments:
+            user (User)
+            course_key (CourseLocator)
+            block_key (BlockUsageLocator)
+
+        Returns:
+            list[modules]
+        """
+        return StudentModule.objects.filter(
+            student=user,
+            course_id=course_key,
+            module_state_key=block_key,
+            state__contains='"selected": [['
+        )
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Computes any information for each XBlock that's necessary to execute
+        this transformer's transform method.
+
+        Arguments:
+            block_structure (BlockStructureCollectedData)
+
+        Returns:
+            dict[UsageKey: dict]
+        """
+        block_structure.request_xblock_fields('mode')
+        block_structure.request_xblock_fields('max_count')
+
+        # For each block check if block is library_content.
+        # If library_content add children array to content_library_children field
+        for block_key in block_structure.topological_traversal():
+            xblock = block_structure.get_xblock(block_key)
+            block_structure.set_transformer_block_data(block_key, cls, 'content_library_children', [])
+            if getattr(xblock, 'category', None) == 'library_content':
+                block_structure.set_transformer_block_data(block_key, cls, 'content_library_children', xblock.children)
+
+    def transform(self, user_info, block_structure):
+        """
+        Mutates block_structure and block_data based on the given user_info.
+
+        Arguments:
+            user_info(object)
+            block_structure (BlockStructureCollectedData)
+        """
+        store = modulestore()
```

Comment:
What will it require to remove the usage of the modulestore in the transform phase? All modulestore-dependent code should reside only in the collect phase. This will allow us to eventually separate LMS (cached-read-only-courses) and Studio (read-write-courses).